better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1129 @@
+make sense of the code directly in front of them, without going on a
+journey discovering how the code works.”
+
+So in our example, what we know about `sort` allow us to reason about
+its use without looking at its implementation.  In fact, local
+reasoning is so fundamental that most of our programming best
+practices exist just to support our ability to do it.  It's why we
+use access control, why we break programs into components like
+functions, types, and modules, and why we try to keep them small.
+
+## Components and Abstraction
+
+Obviously, “knowing something about `sort`” is only possible and
+relevant because `sort` is a separate component with its own name. If
+we had injected the sorting implementation directly into our code,
+we'd have to reason about all of its details, and we'd need a comment
+to tell us that the implementation was sorting the elements.
+
+It's a common fallacy that factoring out a component only makes
+sense if it's used more than once. The real criterion is whether you
+can identify a unit of *abstraction*—the third ingredient of Better
+Code.  If you can think of a simple and descriptive name for
+something, that's usually a good indicator that there's an abstraction
+waiting to be discovered.
+
+In fact, creating a component directly creates the opportunity for an
+additional use site: in testing.  The more code you test separately,
+the better it is for correctness.  The more you can use understandable
+abstractions, the simpler and clearer the use-sites become, which is
+better for correctness. Abstraction and correctness support one
+another in a virtuous cycle.  And abstraction is necessary for
+programming at scale: because we can't keep the whole program in our
+heads, we do what humans always do when faced with complexity: we
+break complicated problems into parts that can be understood in
+isolation.
+
+## Hoare Logic
+
+Programming by contract grew out of Hoare Logic, developed in 1969 by
+the British computer scientist and logician Tony Hoare.  While we
+won't be using Hoare Logic directly, we'll discuss it briefly to
+provide historical context, and because it is the source of crucial
+terminology.
+
+### Preconditions and Postconditions
+
+Hoare used this notation, called a “Hoare triple,”
+
+> {P}S{Q}
+
+which is an assertion that if **precondition** *P* is met, operation
+*S* establishes **postcondition** *Q*.
+
+For example:
+
+- if we start with `x == 2` (precondition), after `x += 1`, `x == 3` (postcondition):
+
+  > {x == 2}x+=1{x == 3}
+
+- if `x` is less than the maximum integer (precondition), after `x
+  += 1`, `x` is greater than the minimum integer (postcondition):
+
+  > {x < Int.max}x+=1{x > Int.min}
+
+What makes preconditions and postconditions useful for formal proofs
+is this composition rule:
+
+> {P}S{Q} ∧ {Q}T{R} ⇒ {P}S;T{R}
+
+Given two valid Hoare triples, if the postconditions of the first are
+preconditions of the second, we can form a new valid triple describing
```

Comment:
Surely, we mean,

```suggestion
Given two valid Hoare triples, if the postconditions of the first are
a superset of the preconditions of the second, we can form a new 
valid triple describing
```

Reply:
No, I mean what I wrote.  You could get there by saying that if {P}S{Q} ⇒ {P}S{Q'} for any Q'⊂Q, which is a subset of the [consequence rule](https://en.wikipedia.org/wiki/Hoare_logic#Consequence_rule).  I am trying to keep the logic stuff as simple as possible and just describing the [rule of composition](https://en.wikipedia.org/wiki/Hoare_logic#Rule_of_composition).  As soon as you bring in the other stuff the visual complexity of the math will start overwhelming the value of its meaning for many people.